product/src/main/java/com/yas/product/service/ProductService.java

```diff
@@ -170,59 +171,69 @@ public ProductListGetVm getProductsWithFilter(int pageNo, int pageSize, String p
-            throw new DuplicatedException(Constants.ErrorCode.GTIN_ALREADY_EXISTED_OR_DUPLICATED, gtin);
-        }
-
-        if (isProductWithSkuAvailable(sku)) {
-            throw new DuplicatedException(Constants.ErrorCode.SKU_ALREADY_EXISTED_OR_DUPLICATED, sku);
+    private void validateExistingProductProperties(HasProductProperties productProperties, Product existingProduct) {
+        checkPropertyExists(productProperties.slug(), existingProduct,
+            productRepository::findBySlugAndIsPublishedTrue, Constants.ErrorCode.SLUG_ALREADY_EXISTED_OR_DUPLICATED);
+        // only check gtin when it's not empty
+        if (StringUtils.isNotEmpty(productProperties.gtin())) {
+            checkPropertyExists(productProperties.gtin(), existingProduct,
+                productRepository::findByGtinAndIsPublishedTrue,
+                Constants.ErrorCode.GTIN_ALREADY_EXISTED_OR_DUPLICATED);
         }
+        checkPropertyExists(productProperties.sku(), existingProduct,
+            productRepository::findBySkuAndIsPublishedTrue, Constants.ErrorCode.SKU_ALREADY_EXISTED_OR_DUPLICATED);
     }
 
-    private void validateProductVariationDuplicates(List<ProductVariationPostVm> variations) {
-        Set<String> seenSlugs = new HashSet<>();
-        Set<String> seenSkus = new HashSet<>();
-        Set<String> seenGtins = new HashSet<>();
-        for (ProductVariationPostVm variation : variations) {
+    private void validateProductVariationDuplicates(ProductSaveVm productSaveVm) {
+        Set<String> seenSlugs = new HashSet<>(Collections.singletonList(productSaveVm.slug()));
+        Set<String> seenSkus = new HashSet<>(Collections.singletonList(productSaveVm.sku()));
+        Set<String> seenGtins = new HashSet<>(Collections.singletonList(productSaveVm.gtin()));
+        for (HasProductProperties variation : productSaveVm.variations()) {
             if (!seenSlugs.add(variation.slug())) {
                 throw new DuplicatedException(Constants.ErrorCode.SLUG_ALREADY_EXISTED_OR_DUPLICATED, variation.slug());
             }
-
-            if (!seenGtins.add(variation.gtin())) {
+            // only check gtin when it's not empty
+            if (StringUtils.isNotEmpty(variation.gtin()) && !seenGtins.add(variation.gtin())) {
                 throw new DuplicatedException(Constants.ErrorCode.GTIN_ALREADY_EXISTED_OR_DUPLICATED, variation.gtin());
             }
-
             if (!seenSkus.add(variation.sku())) {
                 throw new DuplicatedException(Constants.ErrorCode.SKU_ALREADY_EXISTED_OR_DUPLICATED, variation.sku());
             }
         }
     }
 
+    private void validateProductVm(ProductSaveVm productSaveVm, Product existingProduct) {
+        // validate the properties of the main product
+        validateExistingProductProperties(productSaveVm, existingProduct);
+
+        // validate whether the product and its variations contain duplicate properties
+        validateProductVariationDuplicates(productSaveVm);
```

Comment:
you should consider removing this validate because it is validated in line 227 for variants